### ISSUE TITLE
Thoughts on HCCString.h

### ISSUE THREAD

**Issue by CodeVisio** [Harlinn/Harlinn.Windows#1]
Hi,

The `explicit` keyword on `BasicString `constructors on line 956
doesn't allow code like this (unless being pedantic):

```
#define FOO L"foo"
constexpr Core::WideString::CharType * foo = L"foo";

void Bar( const Core::WideString & s ) {...}

later...
{
   Bar( FOO );  // flagged as error, fixed (pedantic) as Bar( Core::WideString( FOO ) );
   Bar( foo ); // flagged as error, fixed (pedantic) as Bar( Core::WideString( foo ) );
}

```
Constructor at 961 suffers the same problem (imagine `Bar( { FOO, _countof( FOO ) } ) `)
Compiled with VS2022 community and C++ 20.

Is that by design?

Thanks

Edited

**Comment by Harlinn** [comment]
Yes, for now, it's by design. This constructor wasn't explicit to begin with, but ended up in an implicit conversion, causing a bug.
Making it explicit, fixed the problem.

Best regards
Espen Harlinn

**Comment by CodeVisio** [comment]
Thank you for your answer.

I've skimmed through the class code (HCCString) and I've noticed some subtleties
but it could be my non-familiar with the class design.
Without building fancy examples you can see one of the subtleties just using the following code:


```
WideString s;
s.Insert( 0, 1, L'H' );

```
This should crash.
That's because, following your code, the const iterator is constructed when the inner reference data is not yet allocated, then memove references a null pointer.

Best Regards

**Comment by Harlinn** [comment]
Thanks for locating this embarrassing bug, I've reimplemented Insert and added a few unit tests. It was, unfortunately, not covered by the tests.  If you find anything else I would certainly like to know :-)

Best regards
Espen Harlinn

**Comment by CodeVisio** [comment]
Glad I could help.

Best Regards

**Comment by Harlinn** [comment]
Have removed `explicit` from most of the `BasicString`  constructors ...
